### ISSUE TITLE
How to limit message batch size or ack in middle of handling message batch?

### ISSUE THREAD

**Issue by mindreader** [kafka4beam/brod#568]
We have a `brod_group_subscriber` that receives messages, usually 5 or 10, but sometimes there are 5000+, which is too many for us to handle at once. So we take the messages and we chunk into blocks of say 500 messages and process each separately, where each block can take 20+ seconds but its fine.

The problem is that when there is an error near the end of this block of messages, it will then reprocess all 5000 messages again from the beginning, which is a long time, and a lot of reprocessing.

So I tried, in my `handle_messages` callback, to ack the offset of the final message in each block of 500, but it doesn't work. Because `handle_messages` is a call, and acking is a cast to the same gen_server, the acks do not get processed until after all 5000 messages are processed, so any exception will guarantee none of those acks get forwarded to the `brod_group_coordinator`. I cannot see a way to get the pid of the coordinator to ack to it directly.

Is there any way to limit the size of message batches or to ack in the middle of processing a single batch?

I've tried some things like limiting the size of messages received through consumer options, but it is fiddly having to estimate average message size and it incredibly slows down processing as a whole.

**Comment by zmstone** [comment]
Hi @mindreader 
`brod_group_subscriber_v2` allows the callback to return `ack` for each message.
https://github.com/kafka4beam/brod/blob/6966f0244b87904363cb8053ac0364ff281411d9/src/brod_group_subscriber_worker.erl#L79-L83

**Comment by mindreader** [comment]
Thanks for the response. I'm not sure this would work because nothing is committed until all 5k messages are processed, so any error prevents any commits.

We've since re-engineered out app to avoid large message batches to get around this, so it isn't a concern any longer.

**Comment by zmstone** [comment]
Maybe you did not observe the commit because the coordinator process collects `ack`s and send a batch to Kaka every 5 seconds.

**Comment by mindreader** [comment]
Sorry if I'm misunderstanding, but looking at the code, it calls my module callback with 5k messages, and then acks or commits the 5000th message, depending on what I return.

But that means I have to fully process all 5000 of them before I can send any indication back to brod that any messages were handled.

Originally I tried to send out of band acks but those acks are targetted to the same genserver that is handling the messages, so it doesn't handle the acks during message processing.

It should be possible to send those out of band acks directly to the supervisor, as that is what the genserver does. It knows the supervisors pid and so it just forwards them along, but my callback module is never given that pid, so it can't send them itself.

The every 5 second commit happens, but it never has a chance to know that I've finished processing any messages until I've processed all of them.
